resources/assets/lib/utils/beatmapset-helper.ts

```diff
@@ -15,3 +22,32 @@ export function nominationsCount(nominations: BeatmapsetNominationsInterface, ty
 export function showVisual(beatmapset: BeatmapsetJson) {
   return !beatmapset.nsfw || (currentUser?.user_preferences?.beatmapset_show_nsfw ?? false);
 }
+
+export function toggleFavourite(beatmapset: BeatmapsetJson) {
+  const retryCallback = () => toggleFavourite(beatmapset);
+
+  if (core.userLogin.showIfGuest(retryCallback)) return;
+
+  const add = !beatmapset.has_favourited;
+
+  // fake immediate change
+  beatmapset.has_favourited = add;
+  beatmapset.favourite_count += add ? 1 : -1;
+
+  $.ajax(route('beatmapsets.favourites.store', { beatmapset: beatmapset.id }), {
+    data: {
+      action: add ? 'favourite' : 'unfavourite',
+    },
+    method: 'POST',
+  })
+  .fail((xhr: JQuery.jqXHR, status: string) => {
+    // undo faked change
+    beatmapset.has_favourited = !add;
+    beatmapset.favourite_count += add ? -1 : 1;
+
+    error(xhr, status, retryCallback);
+  })
+  .done((data: FavouriteResponse) => {
+    beatmapset.favourite_count = data.favourite_count;
+  });
```

Comment:
This indent seems fine but eslint is going to complain about it (and the rule itself has many bugs with constructs like this); the easiest way around it seems to be to put the call on the same line like `}).fail` instead.